### PR TITLE
(do not merge) spike: puppet full-integration tests

### PR DESCRIPTION
**Pull request by er0ck**
spike on POC full-integration puppet tests.
⚠️  caution! this WILL alter your system just like running `puppet apply` on your system.  

this loads and enables most of puppet and runs a given manifest.
it needs work:
as shown with the report example, this does not setup reporting, and i'm sure some other stuff. 

this is strangely slow on my osx 1011 laptop.

please comment with thoughts, ways to make it better.  should we do this at all?

benefits of moving tests from beaker to rspec:

- closer to devs
- easier to run
- will be run more often before commit/PR

problems:
- possibly not much faster (in run time)
- might not run exactly the same way as beaker en-SUT (installed)

### CODE REVIEW

**Comment by er0ck** [comment]
reporting may be able to be added with something like:
`Puppet::Transaction::Report.new("apply", nil, Puppet[:environment], transaction_uuid)`

but i could not get it to work the way i'm creating the apply command here. :man_shrugging:

**Comment by er0ck** [comment]
runtime comparison: (rspec on centos7 slice small box) (beaker SUT is rhel7)
```
✓ erict@erict ~/projects/puppet
 (spike/master/do_not_merge-puppet_full_integration_tests *$%=)$ time be rspec --color spec/integration/apply/test1.
rb


Finished in 1.09 seconds (files took 0.62942 seconds to load)
1 example, 0 failures


real    0m2.550s
user    0m2.300s
sys     0m0.229s
```
```
✓ erict@erict ~/projects/puppet/acceptance
 (spike/master/do_not_merge-puppet_full_integration_tests *$%=)$ be bea^Cr --preserve-hosts --hosts log/latest/hosts
_preserved.yml --color --tests tests/lookup/eyaml_backend.rb
 (spike/master/do_not_merge-puppet_full_integration_tests *$%=)$ time be beaker --preserve-hosts --hosts log/latest/hosts_preserved.yml --color --tests tests/apply/notify.rb

[...]
vpaqbz2hj3xii6q.delivery.puppetlabs.net (vpaqbz2hj3xii6q.delivery.puppetlabs.net) 23:47:19$ puppet apply --verbose /var/folders/zz/zyxvpxvq6csfxvn_n0000000000000/T/apply_manifest.pp.XXXXXX.dTLKWW4w
  Notice: Compiled catalog for vpaqbz2hj3xii6q.delivery.puppetlabs.net in environment production in 0.06 seconds
  Info: Applying configuration version '1497397643'
  Notice: it werks!
  Notice: /Stage[main]/Main/Notify[it werks!]/message: defined 'message' as 'it werks!'
  Notice: Applied catalog in 0.04 seconds

vpaqbz2hj3xii6q.delivery.puppetlabs.net (vpaqbz2hj3xii6q.delivery.puppetlabs.net) executed in 3.96 seconds
Begin teardown
End teardown
tests/apply/notify.rb passed in 3.99 seconds
      Test Suite: tests @ 2017-06-13 23:47:19 +0000

      - Host Configuration Summary -


              - Test Case Summary for suite 'tests' -
       Total Suite Time: 3.99 seconds
      Average Test Time: 3.99 seconds
              Attempted: 1
                 Passed: 1
                 Failed: 0
                Errored: 0
                Skipped: 0
                Pending: 0
                  Total: 1

      - Specific Test Case Status -

Failed Tests Cases:
Errored Tests Cases:
Skipped Tests Cases:
Pending Tests Cases:

No tests to run for suite 'post_suite'
No tests to run for suite 'pre_cleanup'

You can reproduce this run with:
/home/erict/.rvm/gems/ruby-2.3.1/bin/beaker --preserve-hosts --hosts log/latest/hosts_preserved.yml --color --tests tests/apply/notify.rb

You can re-run commands against the already provisioned SUT(s) with:
/home/erict/.rvm/gems/ruby-2.3.1/bin/beaker --preserve-hosts --hosts log/redhat7-64m-osx1011-64a-932bdaf4-5abb-4f30-8523-e60b1aaa0be1.yaml/2017-06-13_23_47_19/hosts_preserved.yml --color --tests tests/apply/notify.rb

Important ENV variables that may have affected your run:
    BUNDLE_BIN          .bundle/bin
    GEM_HOME            /home/erict/.rvm/gems/ruby-2.3.1
    GEM_PATH
    RUBYLIB             /home/erict/.rvm/gems/ruby-2.3.1/gems/bundler-1.14.4/lib
    PATH                /home/erict/.rvm/gems/ruby-2.3.1/bin:/home/erict/.rvm/gems/ruby-2.3.1@global/bin:/home/erict/.rvm/rubies/ruby-2.3.1/bin:/home/erict/.rvm/bin:/opt/local/bin:/opt/local/sbin:/home/erict/bin:/home/erict/bin/local:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin
    home                /home/erict
Beaker completed successfully, thanks.

real    0m6.147s
user    0m1.221s
sys     0m0.159s
```

**Comment by er0ck** [comment]
oops, this needed [skip ci]

**Comment by er0ck** [comment]
⚠️ i just noticed the expects in the spec tests here do nothing, as the described_object is not doing the outputting.  the logger is.  will need to fix

**Comment by er0ck** [comment]
comparison of the report test.

TL;DR: 
rspec: real    0m2.429s
beaker: real    0m5.909s
```
✗ erict@erict ~/projects/puppet
 (spike/master/do_not_merge-puppet_full_integration_tests *$%=)$ time be rspec --color spec/integration/apply/report.rb
Notice: Compiled catalog for erict.rfc1918.puppetlabs.net in environment *root* in 0.02 seconds
Notice: Local environment: 'production' doesn't match server specified environment '*root*', restarting agent run with environment '*root*'
Notice: Compiled catalog for erict.rfc1918.puppetlabs.net in environment *root* in 0.01 seconds
Error: Found 1 dependency cycle:
(Notify[bar] => Notify[foo] => Notify[bar])
Try the '--graph' option and opening the resulting '.dot' file in OmniGraffle or GraphViz
Error: Failed to apply catalog: One or more resource dependency cycles detected in graph


Finished in 1.06 seconds (files took 0.59753 seconds to load)
1 example, 0 failures


real    0m2.429s
user    0m2.153s
sys     0m0.257s
```
```
✗ erict@erict ~/projects/puppet/acceptance
 (spike/master/do_not_merge-puppet_full_integration_tests *$%=)$ time be beaker --preserve-hosts --hosts log/latest/hosts_preserved.yml --color --tests tests/apply/report.rb
Hypervisor for rtpr5vtotfahn2u.delivery.puppetlabs.net is vmpooler

[...]

              - Test Case Summary for suite 'tests' -
       Total Suite Time: 3.77 seconds
      Average Test Time: 3.77 seconds
              Attempted: 1
                 Passed: 0
                 Failed: 0
                Errored: 1
                Skipped: 0
                Pending: 0
                  Total: 1

      - Specific Test Case Status -

Failed Tests Cases:
Errored Tests Cases:
  Test Case tests/apply/report.rb reported: #<Beaker::Host::CommandFailure: Host 'vpaqbz2hj3xii6q.delivery.puppetlabs.net' exited with 1 running:
 puppet apply --verbose /var/folders/zz/zyxvpxvq6csfxvn_n0000000000000/T/apply_manifest.pp.XXXXXX.NndLVtHE
Last 10 lines of output were:
        Notice: Compiled catalog for vpaqbz2hj3xii6q.delivery.puppetlabs.net in environment production in 0.07 seconds
        Info: Applying configuration version '1497398690'
        Error: Failed to apply catalog: Found 1 dependency cycle:
        (Notify[bar] => Notify[foo] => Notify[bar])
        Try the '--graph' option and opening the resulting '.dot' file in OmniGraffle or GraphViz>
    Test line: tests/apply/report.rb:7:in `block (2 levels) in run_test'
Skipped Tests Cases:
Pending Tests Cases:

Failed: errored in TestSuite: report_and_raise_on_failure
#<RuntimeError: Failed while running the tests suite>
No tests to run for suite 'post_suite'
No tests to run for suite 'pre_cleanup'

You can reproduce this run with:
/home/erict/.rvm/gems/ruby-2.3.1/bin/beaker --preserve-hosts --hosts log/latest/hosts_preserved.yml --color --tests tests/apply/report.rb

You can re-run commands against the already provisioned SUT(s) with:
/home/erict/.rvm/gems/ruby-2.3.1/bin/beaker --preserve-hosts --hosts log/redhat7-64m-osx1011-64a-932bdaf4-5abb-4f30-8523-e60b1aaa0be1.yaml/2017-06-14_00_04_45/hosts_preserved.yml --color --tests tests/apply/report.rb

Important ENV variables that may have affected your run:
    BUNDLE_BIN          .bundle/bin
    GEM_HOME            /home/erict/.rvm/gems/ruby-2.3.1
    GEM_PATH
    RUBYLIB             /home/erict/.rvm/gems/ruby-2.3.1/gems/bundler-1.14.4/lib
    PATH                /home/erict/.rvm/gems/ruby-2.3.1/bin:/home/erict/.rvm/gems/ruby-2.3.1@global/bin:/home/erict/.rvm/rubies/ruby-2.3.1/bin:/home/erict/.rvm/bin:/opt/local/bin:/opt/local/sbin:/home/erict/bin:/home/erict/bin/local
:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin
    home                /home/erict
Failed running the test suite.


real    0m5.909s
user    0m1.121s
sys     0m0.147s
```

**Comment by er0ck** [comment]
@johnduarte @branan @joshcooper FYI

**Comment by MosesMendoza** [comment]
@er0ck is this still actively in development? Currently triaging puppet prs to get the list down. Can you close if not?
